src/ir/from_ast/astconv.rs

```diff
@@ -860,94 +850,79 @@ impl<'ctx, 'prog> BuildCtx<'ctx, 'prog> {
-    }
 }
 
 pub fn transform(ns: ast::Namespace) -> ir::Context {
-    let mut sig_map = SigMap::default();
-    let main_idx = ns.main_idx();
-    let (mut ns, order) = crate::utils::Traversal::from(ns).take();
-
-    // Walk over signatures and build a SigMap
-    for (idx, sig) in ns.signatures().map(|(_, sig)| sig).enumerate() {
-        sig_map.insert(sig.name.copy(), Sig::from((sig, idx)));
-    }
+    // creates an empty context with the main index.
+    let mut ctx = ir::Context {
+        entrypoint: ns
+            .main_idx()
+            // index main component after all externals
+            .map(|idx| crate::utils::Idx::new(ns.externals().count() + idx)),
+        ..Default::default()
+    };
+
+    // Walk over signatures and compile signatures to build a SigMap
+    let (mut builders, sig_map): (Vec<_>, SigMap) = ns
+        // pull signatures out of externals
+        .externs
+        .into_iter()
+        .flat_map(|(name, comps)| {
+            comps.into_iter().map(move |comp| (name.clone(), comp))
+        })
+        .map(|(name, sig)| (Some(name), sig, None))
+        // add signatures of components as well as their command bodies
+        .chain(
+            ns.components
+                .into_iter()
+                .map(|comp| (None, comp.sig, Some(comp.body))),
+        )
+        .enumerate()
+        .map(|(idx, (file, sig, body))| {
+            let idx = ir::CompIdx::new(idx);
+            let mut builder = BuildCtx::new(ir::Component::new(body.is_none()));
+            // enable source information saving if this is main or an external.
+            if body.is_none() || Some(idx) == ctx.entrypoint {
+                builder.comp.src_info = Some(InterfaceSrc::new(sig.name.copy()))
+            }
+            // add the file to the externals map if it exists
+            if let Some(file) = file {
+                ctx.externals.entry(file).or_default().push(idx);
+            }
 
-    let mut ctx = ir::Context::default();
-    for (file, exts) in ns.externs {
-        for ext in exts {
-            let idx = sig_map.get(&ext.name).unwrap().idx;
-            log::debug!("Converting external {}: {}", ext.name, idx);
-            let ir_ext = BuildCtx::external(&ctx, ext);
-            ctx.comps.checked_add(idx, ir_ext);
-            ctx.externals.entry(file.clone()).or_default().push(idx);
-        }
-    }
+            // compile the signature
+            builder.comp.cmds = builder.sig(&sig);
 
-    // TODO: Need to handle recursive components as well as mutually recursive components by adding signature ports first in declaration, so that foreign keys can be resolved
-    // declare all components in the proper order
-    for (cidx, comp) in ns.components.iter().enumerate() {
-        let idx = sig_map.get(&comp.sig.name).unwrap().idx;
+            let irsig = Sig::new(idx, &builder.comp, &sig);
+            ((idx, builder, body), (sig.name.take(), irsig))
+        })
+        .unzip();
```

Comment:
This chain of operations is too long and hard to understand. Can we split it up into two before we do the `enumerate` call?